### PR TITLE
Add "empty_object_spaces" to braces_spacing

### PR DESCRIPTION
**Pull request by gerry3**

### CODE REVIEW

**Comment by gerry3** [comment]
cc @charlierudolph

**Comment by gerry3** [comment]
@AsaAyers @charlierudolph thoughts?

File: src/rules/braces_spacing.coffee [hunk 1]
```diff
@@ -55,7 +65,10 @@ module.exports = class BracesSpacing
 
         return null unless @tokensOnSameLine firstToken, secondToken
 
-        expected = tokenApi.config[@rule.name].spaces
+        isEmpty = firstToken[0] is '{' and secondToken[0] is '}'
+        expected = tokenApi.config[@rule.name].emptySpaces if isEmpty
+
+        expected = tokenApi.config[@rule.name].spaces unless expected?
```
Comment (charlierudolph) [inline src/rules/braces_spacing.coffee]: How do you feel about a `getExpectedSpaces` method

```coffee
getExpectedSpaces: (tokenApi, firstToken, secondToken) ->
  {emptySpaces, spaces} = tokenApi.config[@rule.name]
  if firstToken[0] is '{' and secondToken[0] is '}'
    emptySpaces ? spaces
  else
    spaces
```

File: test/test_braces_spacing.coffee [hunk 2]
```diff
@@ -27,15 +27,23 @@ sources =
...
+        oneSpace: '{ }'
+        twoSpaces: '{  }'
     stringInterpolation:
         noSpaces: "\"\#{foo}\""
         oneSpace: "\"\#{ foo }\""
 
 configs =
     zeroSpaces:
-        braces_spacing: {level: 'error', spaces: 0}
+        braces_spacing: {level: 'error', spaces: 0, emptySpaces: undefined}
```
Comment (charlierudolph) [inline test/test_braces_spacing.coffee]: I vote for just leaving off emptySpaces.

File: test/test_braces_spacing.coffee [hunk 3]
```diff
@@ -27,15 +27,23 @@ sources =
...
     oneSpace:
-        braces_spacing: {level: 'error', spaces: 1}
+        braces_spacing: {level: 'error', spaces: 1, emptySpaces: undefined}
+    zeroEmptySpaces:
```
Comment (charlierudolph) [inline test/test_braces_spacing.coffee]: oneSpaceZeroEmptySpaces

File: test/test_braces_spacing.coffee [hunk 4]
```diff
@@ -156,4 +196,38 @@ vows.describe('braces_spacing').addBatch({
                            configs.oneSpace)
 
 
+    'enabled with empty spaces set to 0':
```
Comment (charlierudolph) [inline test/test_braces_spacing.coffee]: enabled with spaces set to 1 and empty spaces set to 0

File: test/test_braces_spacing.coffee [hunk 5]
```diff
@@ -156,4 +196,38 @@ vows.describe('braces_spacing').addBatch({
...
+                        'There should be 0 spaces inside "}"'])
+
+        'two spaces inside on both braces':
+            shouldFail(sources.empty.twoSpaces,
+                       configs.zeroEmptySpaces,
+                       ['There should be 0 spaces inside "{"',
+                        'There should be 0 spaces inside "}"'])
+
+
+    'enabled with empty spaces set to 1':
```
Comment (charlierudolph) [inline test/test_braces_spacing.coffee]: enabled with spaces set to zero and empty spaces set to 1

File: test/test_braces_spacing.coffee [hunk 6]
```diff
@@ -27,15 +27,23 @@ sources =
...
 configs =
     zeroSpaces:
-        braces_spacing: {level: 'error', spaces: 0}
+        braces_spacing: {level: 'error', spaces: 0, emptySpaces: undefined}
     oneSpace:
-        braces_spacing: {level: 'error', spaces: 1}
+        braces_spacing: {level: 'error', spaces: 1, emptySpaces: undefined}
+    zeroEmptySpaces:
+        braces_spacing: {level: 'error', spaces: 1, emptySpaces: 0}
+    oneEmptySpace:
```
Comment (charlierudolph) [inline test/test_braces_spacing.coffee]: zeroSpacesOneEmptySpace

File: test/test_braces_spacing.coffee [hunk 7]
```diff
@@ -27,15 +27,23 @@ sources =
                    {  foo,
                       bar  } = x
                    '''
+    empty:
+        noSpaces: '{}'
+        oneSpace: '{ }'
+        twoSpaces: '{  }'
```
Comment (charlierudolph) [inline test/test_braces_spacing.coffee]: I had the keys in alphabetical order. Can you update the placement to keep it in alpha order? Same for the configs below

**Comment by swang** [comment]
1. Anything that is written into the config.json should follow snake_case formatting, so it should by empty_spaces. Even better is probably empty_object_spacing. Bit longer but I see empty spaces and  I have no idea what that means.

2.  The way this test is written with all the source code makes it very unclear what each test is testing. When one test fails, I'm going to look into this source code around that line, then see a variable in place of source code and will have to scroll/page up to find out which source code is broken. I don't need this change to be in this PR, but a test script shouldn't be written out like this IMO.

**Comment by gerry3** [comment]
@swang I originally went with snake_case, but then discovered that it was more common (2 rules vs 1) to use camelCase and changed it. I will use snake_case for this now, thanks.

Rule `max_line_length` has `limitComments`.
Rule `no_trailing_whitespace` has `allowed_in_comments` & `allowed_in_empty_lines`.
Rule `prefer_english_operator` has `doubleNotLevel`.

**Comment by gerry3** [comment]
@swang @charlierudolph comments addressed

**Comment by gerry3** [comment]
@AsaAyers @charlierudolph @swang thoughts?

**Comment by charlierudolph** [comment]
looks good to me

**Comment by gerry3** [comment]
@AsaAyers can this be merged please?

**Comment by AsaAyers** [comment]
I'm leaving myself a reminder to do a patch release on Monday. I'm going to be busy all weekend and even for patch releases I just don't like to do them when I won't be around.

Thanks for the PR :smile:

**Comment by AsaAyers** [comment]
This is included in 1.9.4

**Pull request merged**